Eagle3 that supports the Minicpm3 model

LDLINGLINGLING opened the pull request:
## Purpose
Eagle3 that supports the Minicpm3 model
## Test Plan
————————————————
200 samples, 10 concurrency, single RTX 4090:​​
Total output tokens: 106,678
Number of draft proposals: 43,221
Number of draft tokens generated: 259,326
Number of accepted tokens: 63,669
Mean acceptance length: 2.47

Acceptance rate at token 0: 0.58
Acceptance rate at token 1: 0.34
Acceptance rate at token 2: 0.22
Acceptance rate at token 3: 0.15
Acceptance rate at token 4: 0.11

Baseline (non-speculative): 215.84 s
Speculative decoding model: 133.24 s
Speedup: 1.62×

​20 samples, single concurrency, single RTX 4090:​​
Total output tokens: 21,690
Number of draft proposals: 8,914
Number of draft tokens generated: 53,484
Number of accepted tokens: 12,814
Mean acceptance length: 2.44

Acceptance rate at token 0: 0.57
Acceptance rate at token 1: 0.34
Acceptance rate at token 2: 0.21
Acceptance rate at token 3: 0.14
Acceptance rate at token 4: 0.10

Baseline (non-speculative): 382.29 s
Speculative decoding model: 224.20 s
Speedup: 1.71×
Average acceptance length: 2.44

​100 samples, 5 concurrency, single RTX 4090:​​
Total output tokens: 103,920
Number of draft proposals: 42,227
Number of draft tokens generated: 253,362
Number of accepted tokens: 61,896
Mean acceptance length: 2.47

Acceptance rate at token 0: 0.58
Acceptance rate at token 1: 0.34
Acceptance rate at token 2: 0.22
Acceptance rate at token 3: 0.15
Acceptance rate at token 4: 0.10

Baseline (non-speculative): 412.49 s
Speculative decoding model: 238.08 s
Speedup: 1.73×
Average acceptance length: 2.47
## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

Commit:
增加了minicpm的eagle3投机解码函数

LDLINGLINGLING committed:
use English note and add auxiliary functions for Eagle3

LDLINGLINGLING committed:
Merge branch 'main' into main

youkaichao commented:
cc @luccafong @benchislett can you help take a look?

benchislett commented on vllm/model_executor/models/minicpm.py:
```diff
@@ -563,16 +577,36 @@ def _init_model(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        model_output = self.model(input_ids, positions, intermediate_tensors,
+                                  inputs_embeds)
+        
+        # Check if aux hidden states are returned (consistent with LLaMA's handling approach)
```
> Check if aux hidden states are returned (consistent with LLaMA's handling approach)
What does this mean? Does "LLaMA's handling approach" refer to the way aux_hidden_states are returned? If so, that's not exclusive to Llama it's just how it is done in vLLM for EAGLE3. This comment can probably be removed.

benchislett commented on vllm/model_executor/models/minicpm.py:
```diff
@@ -563,16 +577,36 @@ def _init_model(self, *, vllm_config: VllmConfig, prefix: str = ""):
         positions: torch.Tensor,
         intermediate_tensors: Optional[IntermediateTensors] = None,
         inputs_embeds: Optional[torch.Tensor] = None,
     ) -> Union[torch.Tensor, IntermediateTensors]:
```
The bot is right

benchislett commented:
Overall I have no issues with this PR besides the nitpicks.

However, we are seeing a steady increase in the support of EAGLE3 for new models. We should discuss (separately to this PR) how we want to handle this build-out, as currently there is a lot of duplicated code in every model which needs to support EAGLE3.

LDLINGLINGLING committed:
The code has been modified according to the official requirements.

1. Removed unnecessary comments
2. Modified the forward return format as required

Commit:
修改了规范，以便通过pre-commit

luccafong reviewed (approved):
lgtm! thanks for adding it

hmellor committed:
Merge branch 'main' into pr/LDLINGLINGLING/24243

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
`pre-commit`

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
Review comment

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

luccafong committed:
Merge branch 'main' into main

LDLINGLINGLING committed:
Merge upstream/main: resolve conflicts in speculative.py - combine eagle3 support for both minicpm and gpt_oss models
